Add a simple cloud provider for e2e tests on kubemark

bskiba opened the pull request:
**What this PR does / why we need it**:
Adds a simplified cloud provider for kubemark. This enables us to add and
remove nodes and operate on nodegroups while running tests on kubemark.

This is needed to run scalability tests for cluster autoscaler on kubemark.
See https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/kubemark_integration.md

**Release note**:
```
NONE
```

bskiba committed:
Add a simple cloud provider for e2e tests on kubemark

This is needed for cluster autoscaler e2e test to
run on kubemark. We need the ability to add and
remove nodes and operate on nodegroups. Kubemark
does not provide this at the moment.

shyamjvs commented:
Bazel failed with:

```
test/e2e/framework/framework.go:46: can't find import: "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
```

It has recently been updated to `clientset "k8s.io/client-go/kubernetes"`. Please change.

shyamjvs commented on test/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,342 @@
+/*
```
I would suggest putting this file at `pkg/kubemark/cloudprovider.go` instead, as we usually don't check-in go code under this path but just have setup scripts and configs. A better place would have been `pkg/cloudproviders/providers/`, but it's being deprecated.

shyamjvs commented on test/e2e/framework/framework.go:
```diff
@@ -190,6 +193,16 @@ func (f *Framework) BeforeEach() {
 		Expect(err).NotTo(HaveOccurred())
 		f.ClientPool = dynamic.NewClientPool(config, api.Registry.RESTMapper(), dynamic.LegacyAPIPathResolverFunc)
+		if ProviderIs("kubemark") && TestContext.CloudConfig.UseKubemarkProvider && TestContext.CloudConfig.KubemarkProvider == nil {
+			kubemarkClient, err := kube_client.NewForConfig(config)
```
s/kube_client/clientset/g

shyamjvs commented on test/e2e/framework/test_context.go:
```diff
@@ -196,6 +201,7 @@ func RegisterCommonFlags() {
 func RegisterClusterFlags() {
 	flag.BoolVar(&TestContext.VerifyServiceAccount, "e2e-verify-service-account", true, "If true tests will verify the service account before running.")
 	flag.StringVar(&TestContext.KubeConfig, clientcmd.RecommendedConfigPathFlag, os.Getenv(clientcmd.RecommendedConfigPathEnvVar), "Path to kubeconfig containing embedded authinfo.")
+	flag.StringVar(&TestContext.ExternalKubeConfig, fmt.Sprintf("%s-%s", clientcmd.RecommendedConfigPathFlag, "external"), clientcmd.RecommendedHomeFile, "Path to kubeconfig containing embedded authinfo for external cluster. Valid only for kubemark")
```
You probably can remove "Valid only for kubemark" line as this possibly can be used even without kubemark (e.g if you want to use a custom kubeconfig). Or maybe just say something like "Used for talking to kubemark master".

shyamjvs commented on test/e2e/framework/test_context.go:
```diff
@@ -154,6 +156,9 @@ type CloudConfig struct {
 
 	Provider cloudprovider.Interface
+
+	UseKubemarkProvider bool
```
Can we avoid this extra flag? I guess we can just use `ExternalKubeConfig` (if specified) to create `KubemarkProvider` when we know that `IsProvider("kubemark") == true`.

bskiba replied:
Currently the ExternalKubeConfig defaults to the kubeconfig in home directory (mostly because it was convenient for local testing) so it is always specified. But removing the default to reduce the number of flags sounds reasonable.

shyamjvs commented on test/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,342 @@
+			client:    externalClient,
+		},
+		kubemarkCluster: KubemarkCluster{
+			nodeLister:        kubemarkInformerFactory.InformerFor(&apiv1.Node{}, newNodeInformer),
```
Since you're getting a normal informer from the factory (unlike the 2 informers above which are namespaced), I guess just doing a `kubemarkInformerFactory.Core().V1().Nodes()` here should be enough.

shyamjvs commented on test/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,342 @@
+	return cache.NewSharedIndexInformer(podListWatch, &apiv1.Pod{}, resyncPeriod, nil)
+}
+
+func newNodeInformer(kubeClient kube_client.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
```
We don't need this logic and can get rid of the function. The informer can directly be obtained from the factory as mentioned in https://github.com/kubernetes/kubernetes/pull/49382/files#r129030754.

shyamjvs commented:
@bskiba Overall the implementation seems quite clean. Thanks for the PR :)
Besides the above comments, it would be nice if you implement kubemark under our generic cloudprovider interface (ref: https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go#L30) for consistency. You can skip some unneeded methods (like LoadBalancer, DNS, Routes, etc), but others seem relevant (like Initialize, Instances, ProviderName, etc). Will make the code possibly cleaner too.

wojtek-t commented:
> Besides the above comments, it would be nice if you implement kubemark under our generic cloudprovider interface

Note that cloudprovider will be moved to a separate repo. So no one will accept a PR adding a new cloud provider now. You may want to unify it with the interface, but making it "kubemark internals" for now is the way to proceed with this kind of change.

shyamjvs commented:
> Note that cloudprovider will be moved to a separate repo. So no one will accept a PR adding a new cloud provider now. You may want to unify it with the interface, but making it "kubemark internals" for now is the way to proceed with this kind of change.

+1, I agree. To clarify, I was asking just to implement the interface (but I wouldn't be very particular about it). For the code I suggested putting this is pkg/kubemark instead of pkg/cloudproviders for the same reason you mentioned (https://github.com/kubernetes/kubernetes/pull/49382#discussion_r129010037).

shyamjvs commented:
Btw - please add the new flags you defined in `hack/verify-flags/known-flags.txt`

bskiba commented:
Pushed fixes for the comments and fixed some lint errors. Also changed the way node template is taken from existing nodes - this does not depend on the name any more (which was not working in case the original hollow node was deleted) but a node-kind label that is present on all hollow nodes. 
Wrt implementing the CloudProvider interface, I was thinking about it but it felt like a bit of an overshot for this change.

shyamjvs commented:
@bskiba Just checked, the failure is unrelated to your PR. It's recent flakiness with gsutil (https://github.com/GoogleCloudPlatform/gsutil/issues/446).

```
I0726 08:21:15.062] Your "OAuth 2.0 Service Account" credentials are invalid. Please run
I0726 08:21:15.063]   $ gcloud auth login
I0726 08:21:15.063] Duplicate type [0:0:2]
I0726 08:21:38.893] CommandException: 1 file/object could not be transferred.
I0726 08:21:38.975] FAILED
```

bskiba commented:
/retest

bskiba commented:
/retest

shyamjvs commented on test/e2e/framework/framework.go:
```diff
@@ -190,6 +192,16 @@ func (f *Framework) BeforeEach() {
+		if ProviderIs("kubemark") && TestContext.ExternalKubeConfig != "" && TestContext.CloudConfig.KubemarkProvider == nil {
+			kubemarkClient, err := clientset.NewForConfig(config)
+			Expect(err).NotTo(HaveOccurred())
+			externalConfig, err := clientcmd.BuildConfigFromFlags("", TestContext.ExternalKubeConfig)
```
Please also set `externalConfig.QPS` and `exnternalConfig.Burst` like we do for `config`.

bskiba replied:
Done.

shyamjvs commented on test/e2e/framework/framework.go:
```diff
@@ -190,6 +192,18 @@ func (f *Framework) BeforeEach() {
+			Expect(err).NotTo(HaveOccurred())
+			externalClient, err := clientset.NewForConfig(externalConfig)
+			Expect(err).NotTo(HaveOccurred())
+			TestContext.CloudConfig.KubemarkProvider, err = kubemark.NewProvider(externalClient, kubemarkClient, make(chan struct{}))
```
This channel doesn't serve any purpose. To make it meaningful, you should have the channel stored as a field inside `Framework` that you set here and later close inside `AfterEach()` method. For e.g - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L84

shyamjvs commented on pkg/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,368 @@
+			if ok {
+				return nodeGroup, nil
+			}
+			return "", fmt.Errorf("can't find nodegroup for node %s. Node exists but does not have the %s label", nodeGroupLabel, node)
```
Maybe make the error msg shorter? Something like "can't find nodegroup for node %s missing the label '%s' "

shyamjvs commented on pkg/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,368 @@
+			return "", fmt.Errorf("can't find nodegroup for node %s. Node exists but does not have the %s label", nodeGroupLabel, node)
+		}
+	}
+	return "", fmt.Errorf("can't find nodegroup for node %s", node)
```
"can't find node %s" ?

shyamjvs commented on pkg/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,368 @@
+
+// GetNodeGroupForNode returns the name of the node group to which node belongs.
+// Returns an error if node group for node was not found.
+func (kubemarkProvider *Provider) GetNodeGroupForNode(node string) (string, error) {
```
Is this function used anywhere?

shyamjvs commented on pkg/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,368 @@
+	return fmt.Errorf("can't delete node %s from nodegroup %s. Node does not exist", node, nodeGroup)
+}
+
+func (kubemarkProvider *Provider) getNodeGroupByName(nodeGroup string) *apiv1.ReplicationController {
```
Is this function used anywhere? If no, please remove.

shyamjvs commented on pkg/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,368 @@
+	nodesToDeleteLock sync.Mutex
+}
+
+// NewProvider creates Provider using the privided clients to talk to external
```
s/privided/provided/

shyamjvs commented on pkg/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,368 @@
+
+func (kubemarkProvider *Provider) addNodesToNodeGroup(nodeGroup string, delta int) error {
+	if delta < 0 {
+		return fmt.Errorf("delta has to be positive. Got %d", delta)
```
nit: non-negative

shyamjvs commented on pkg/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,368 @@
+
+func (kubemarkProvider *Provider) removeNodesFromNodeGroup(nodeGroup string, delta int) error {
+	if delta < 0 {
+		return fmt.Errorf("delta has to be positive. Got %d", delta)
```
nit: non-negative

shyamjvs commented on pkg/kubemark/cloudprovider.go:
```diff
@@ -0,0 +1,368 @@
+	return nil
+}
+
+func (kubemarkProvider *Provider) addNodeToNodeGroup(nodeGroup string) error {
```
nit: Could you move this function to below addNodesToNodeGroup just for better readability.

shyamjvs commented:
Looks in a good shape. A few comments on how you're creating the informers and starting them:
- Redefine your struct to store the `Lister` and `HasSynced` functions for each informer as fields in the struct
- You should break the `NewProvider` function into two:
  - First one should initialize all your informers and add event handlers to them
  - Second one should run the controller - i.e it should `WaitForCacheSync(..each informer's HasSynced function..)` and then keep running until you receive from the stop channel (or ofc crash).
- Also I'm more inclined towards naming this as KubemarkController over KubemarkProvider because that's more of what this is
- At a couple of places you're listing from the informer's store and then filtering for pods having a given label (e.g https://github.com/bskiba/kubernetes/blob/3015297e3e9b1350a5b54e85d25006c7ee07865f/pkg/kubemark/cloudprovider.go#L128). Please instead list from the `Lister` directly using a label selector.

For e.g pattern, see - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/nodecontroller.go#L161-L165

shyamjvs commented:
/retest

bskiba commented:
/retest

bskiba commented:
/retest

bskiba commented:
/retest

shyamjvs commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,353 @@
+// Controller is a simplified version of cloud provider for kubemark. It allows
+// to add and delete nodes from a kubemark cluster and introduces nodegroups
+// by applying labels to the kubemark's hollow-nodes.
+type Controller struct {
```
Please rename this to KubemarkController. Just not to overload the name we already use for our client-go controller.

bskiba replied:
Done

shyamjvs commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,353 @@
+// by manipulating the replication controllers.
+type externalCluster struct {
+	rcLister        cache.SharedInformer
+	rcListerSynced  cache.InformerSynced
```
s/rcListerSynced/rcSynced (it's not the Lister that is synced, but the informer)
Also make the similar change elsewhere.

bskiba replied:
Done.

shyamjvs commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,353 @@
+		if !ok {
+			continue
+		}
+		if pod.ObjectMeta.Labels[nodeGroupLabel] == nodeGroup {
```
No need to List() all the nodes from the store and then manually filter on a given label. You can directly do `podLister.List(<label-selector>)`
For e.g - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/gc_controller_test.go#L203

shyamjvs commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,353 @@
+	size := 0
+	for _, rcObj := range kubemarkController.externalCluster.rcLister.GetStore().List() {
+		rc := rcObj.(*apiv1.ReplicationController)
+		if rc.Labels[nodeGroupLabel] == nodeGroup {
```
Same as above.

shyamjvs commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,353 @@
+}
+
+func (kubemarkController *Controller) removeNodeFromNodeGroup(nodeGroup string, node string) error {
+	for _, podObj := range kubemarkController.externalCluster.podLister.GetStore().List() {
```
Also here can you directly use label-selector?

shyamjvs commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,353 @@
+}
+
+func (kubemarkController *Controller) getReplicationControllerByName(name string) *apiv1.ReplicationController {
+	for _, obj := range kubemarkController.externalCluster.rcLister.GetStore().List() {
```
Also here - you can use fieldSelector.

shyamjvs commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,353 @@
+}
+
+func (kubemarkController *Controller) getNodeNameForPod(podName string) (string, error) {
+	for _, obj := range kubemarkController.externalCluster.podLister.GetStore().List() {
```
Here too.

bskiba commented:
/retest

gmarek commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,345 @@
+		UpdateFunc: controller.kubemarkCluster.removeUnneededNodes,
+	})
+
+	rand.Seed(time.Now().UTC().UnixNano())
```
This seeds global generator, which can be seeded somewhere else by something. You can use your private `rand.Rand` generator, store it in the object and seed it here.

gmarek commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,345 @@
+}
+
+// GetNodesForNodegroup returns list of the nodes in the node group.
+func (kubemarkController *KubemarkController) GetNodesForNodegroup(nodeGroup string) ([]string, error) {
```
GetNodeNamesForNodeGroup or return slice of Nodes

gmarek commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,345 @@
+	return nil
+}
+
+func (kubemarkController *KubemarkController) addNodesToNodeGroup(nodeGroup string, delta int) error {
```
Why not have a single 'updateNodeGroupSize' with delta that can be negative?

gmarek commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,345 @@
+	node.Name = fmt.Sprintf("%s-%d", nodeGroup, rand.Int63())
+	node.Labels = map[string]string{nodeGroupLabel: nodeGroup, "name": node.Name}
+	node.Spec.Template.Labels = node.Labels
+	_, err = kubemarkController.externalCluster.client.CoreV1().ReplicationControllers(node.Namespace).Create(node)
```
You probably want to add retries here.

gmarek commented on pkg/kubemark/controller.go:
```diff
@@ -0,0 +1,345 @@
+				return fmt.Errorf("can't delete node %s from nodegroup %s. Node is not in nodegroup", node, nodeGroup)
+			}
+			policy := metav1.DeletePropagationForeground
+			err := kubemarkController.externalCluster.client.CoreV1().ReplicationControllers(namespaceKubemark).Delete(
```
Retry this as well.

gmarek commented on test/e2e/framework/test_context.go:
```diff
@@ -201,6 +204,7 @@ func RegisterCommonFlags() {
 func RegisterClusterFlags() {
 	flag.BoolVar(&TestContext.VerifyServiceAccount, "e2e-verify-service-account", true, "If true tests will verify the service account before running.")
 	flag.StringVar(&TestContext.KubeConfig, clientcmd.RecommendedConfigPathFlag, os.Getenv(clientcmd.RecommendedConfigPathEnvVar), "Path to kubeconfig containing embedded authinfo.")
+	flag.StringVar(&TestContext.ExternalKubeConfig, fmt.Sprintf("%s-%s", "external", clientcmd.RecommendedConfigPathFlag), "", "Path to kubeconfig containing embedded authinfo for external cluster.")
```
s/ExternalKubeConfig/KubemarkExternalKubeConfig/ or something - we need to make it explicit that it's a Kubmark-only thing.

gmarek commented:
Just few nits.

bskiba commented:
Addressed the comments, PTAL.

bskiba commented:
/test pull-kubernetes-verify
--

bskiba commented:
/test pull-kubernetes-verify

shyamjvs commented:
/lgtm

gmarek commented:
Please squash comments.

shyamjvs commented:
/lgtm

shyamjvs commented:
@gmarek Seems like your approval should be enough now :)

gmarek commented:
/approve

bskiba commented:
/test pull-kubernetes-verify

bskiba commented:
/test pull-kubernetes-verify

bskiba commented:
/test pull-kubernetes-verify

gmarek commented:
/retest

gmarek commented:
Sorry, missclick.

bskiba commented:
/retest

gmarek commented:
/lgtm

shyamjvs commented:
/retest-not-required

bskiba commented:
/retest

shyamjvs commented:
/joke
